### PR TITLE
circuits: zk-circuits: `VALID WALLET CREATE`: Implement refactored circuit

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR adds in the new definition of the `VALID WALLET CREATE` circuit to the refactor. This circuit validates that:
1. The commitment to the private wallet shares is correctly computed. This commitment is given as a public variable.
2. The underlying wallet, constructed as a combination of secret shares, is a valid, zero'd out wallet.

### Testing
- Unit tests pass, tested cases for cheating prover on invalid commitment, non-zero order, and non-zero balance.

### CODE REVIEW

File: circuits/src/zk_circuits/valid_wallet_create.rs [hunk 1]
```diff
@@ -222,11 +180,50 @@ impl<const MAX_FEES: usize> CommitVerifier for ValidWalletCreateCommitment<MAX_F
...
+{
+    type VarType = ValidWalletCreateStatementVar<MAX_BALANCES, MAX_ORDERS, MAX_FEES>;
+    type ErrorType = (); // Does not error
+
+    fn commit_public<CS: RandomizableConstraintSystem>(
+        &self,
+        cs: &mut CS,
+    ) -> Result<Self::VarType, Self::ErrorType> {
+        let private_commitment_var = self.private_shares_commitment.commit_public(cs).unwrap();
+        let public_shares_var = self.public_wallet_shares.commit_public(cs).unwrap();
```
Comment (akirillo) [inline circuits/src/zk_circuits/valid_wallet_create.rs:220]: Is it safe to be unwrapping here (and in the other commitment traits)? prob ok for now to get this out the door anyway

File: circuits/src/zk_circuits/valid_wallet_create.rs [hunk 2]
```diff
@@ -301,107 +286,136 @@ where
...
+        assert_invalid_witness_statement(witness, statement);
+    }
+
+    /// Tests the case in which a non-zero order is given
+    #[test]
+    fn test_nonzero_order() {
+        let mut wallet = create_empty_wallet();
+        wallet.orders[0] = INITIAL_ORDERS[0].clone();
+
+        let (witness, statement) = create_witness_statement_from_wallet(&wallet);
```
Comment (akirillo) [inline circuits/src/zk_circuits/valid_wallet_create.rs:408]: i may be missing something obvious, but why should this test (and the one below) fail? yeah, we fudge the orders to be non-zero, but we create a witness statement from the wallet after having done this, and afaict `create_witness_statement_from_wallet` makes no assumptions about the wallet thats passed in being an empty wallet.

perhaps in this line we should call `create_default_witness_statment`?

**Review by akirillo — COMMENTED** [review]
had some questions, but assuming tests pass lgtm!
interfaces looking clean

**Comment by akirillo** [comment]
Looks like you're getting some build errors, too?

File: circuits/src/zk_circuits/valid_wallet_create.rs (hunk 1 above)
Comment (joeykraut) [reply circuits/src/zk_circuits/valid_wallet_create.rs:220]: Most of these commitment traits do not error, the error is allowed for generics, but this `unwrap` and most others are safe

File: circuits/src/zk_circuits/mod.rs [hunk 3]
```diff
@@ -96,11 +102,142 @@ mod test_helpers {
...
+        let mut wallet2 = SizedWalletShare {
+            balances: balances2.try_into().unwrap(),
+            orders: orders2.try_into().unwrap(),
+            fees: fees2.try_into().unwrap(),
+            keys: keychain2,
+            blinder: wallet.blinder - blinder_share,
+        };
+
+        // Blind the shares
+        wallet1.blind();
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_circuits/mod.rs:235]: > Looks like you're getting some build errors, too?

Yeah, CI will fail for a while during this refactor. In fact the last step of the implementation plan is what will fix CI (integrate the updated circuits into `core`). This is why I'm targeting an integration branch; to keep `main` green.

**Comment by joeykraut** [comment]
> 

Yeah, CI will fail for a while during this refactor. In fact the last step of the implementation plan is what will fix CI (integrate the updated circuits into core). This is why I'm targeting an integration branch; to keep main green.

File: circuits/src/zk_circuits/valid_wallet_create.rs (hunk 2 above)
Comment (joeykraut) [reply circuits/src/zk_circuits/valid_wallet_create.rs:408]: We want `create_witness_statement_from_wallet` to create an invalid witness; the fact that it makes no assumptions *is* the desired behavior. This test case is testing that the constraints are not satisfied for an invalid witness; and `create_default_witness_statement` creates a *valid* witness.

File: circuits/src/zk_circuits/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_circuits/mod.rs:235]: You're correct, changed this for testing, forgot to change back, good catch!

File: circuits/src/zk_circuits/valid_wallet_create.rs [hunk 4]
```diff
@@ -57,66 +47,54 @@ pub struct ValidWalletCreate<
...
-        // Check that the commitment is to an empty wallet with the given randomness
-        // keys, and fees
-        Self::check_commitment(cs, expected_commit, witness)?;
+        // Validate the commitment given in the statement is a valid commitment to the private secret shares
+        let commitment =
+            WalletShareCommitGadget::compute_commitment(&witness.private_wallet_share, cs)?;
+        cs.constrain(commitment - statement.private_shares_commitment);
+
+        // Unblind the shares then reconstruct the wallet
+        witness.private_wallet_share.unblind();
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_circuits/valid_wallet_create.rs:68]: See above
